lib/models/os.ts

```diff
@@ -0,0 +1,674 @@
+			deviceTypesUtil.getBySlug(types, deviceTypeSlug),
+		);
+
+	/**
+	 * @summary Get OS versions download size
+	 * @description Utility method exported for testability.
+	 * @name _getDownloadSize
+	 * @private
+	 * @function
+	 * @memberof balena.models.os
+	 */
+	const _getDownloadSize = withDeviceTypesEndpointCaching(
+		(deviceType: string, version: string) =>
+			request
+				.send({
+					method: 'GET',
+					url: `/device-types/v1/${deviceType}/images/${version}/download-size`,
+					baseUrl: apiUrl,
+				})
+				.get('body')
+				.get('size'),
+	);
+
+	/**
+	 * @summary Get OS versions
+	 * @description Utility method exported for testability.
+	 * @name _getOsVersions
+	 * @private
+	 * @function
+	 * @memberof balena.models.os
+	 */
+	const _getOsVersions = withDeviceTypesEndpointCaching((deviceType: string) =>
+		request
+			.send({
+				method: 'GET',
+				url: `/device-types/v1/${deviceType}/images`,
+				baseUrl: apiUrl,
+			})
+			.get('body')
+			.then(({ versions, latest }) => {
+				versions.sort(bSemver.rcompare);
+				const potentialRecommendedVersions = reject(
+					versions,
+					version =>
+						semver.prerelease(version) || isDevelopmentVersion(version),
+				);
+				const recommended =
+					(potentialRecommendedVersions != null
+						? potentialRecommendedVersions[0]
+						: undefined) || null;
+
+				return {
+					versions,
+					recommended,
+					latest,
+					default: recommended || latest,
+				};
+			}),
+	);
+
+	/**
+	 * @summary Clears the cached results from the `device-types/v1` endpoint.
+	 * @description Utility method exported for testability.
+	 * @name _clearDeviceTypesEndpointCaches
+	 * @private
+	 * @function
+	 * @memberof balena.models.os
+	 */
+	const _clearDeviceTypesEndpointCaches = () => {
+		_getDeviceTypes.clear();
+		_getDownloadSize.clear();
+		return _getOsVersions.clear();
+	};
+
+	const normalizeVersion = (v: string) => {
+		if (!v) {
+			throw new Error(`Invalid version: ${v}`);
+		}
+		if (v === 'latest') {
+			return v;
+		}
+		const vNormalized = v[0] === 'v' ? v.substring(1) : v;
+		if (!BALENAOS_VERSION_REGEX.test(vNormalized)) {
+			throw new Error(`Invalid semver version: ${v}`);
+		}
+		return vNormalized;
+	};
+
+	const deviceImageUrl = (deviceType: string, version: string) =>
+		`/download?deviceType=${deviceType}&version=${version}`;
+
+	const fixNonSemver = (version: string) => {
+		if (version == null) {
+			return version;
+		}
+
+		return version.replace(/\.rev(\d+)/, '+FIXED-rev$1');
+	};
+
+	const unfixNonSemver = (version: string) => {
+		if (version == null) {
+			return version;
+		}
+
+		return version.replace(/\+FIXED-rev(\d+)/, '.rev$1');
+	};
+
+	/**
+	 * @summary Get the max OS version satisfying the given range.
+	 * @description Utility method exported for testability.
+	 * @name _getMaxSatisfyingVersion
+	 * @private
+	 * @function
+	 * @memberof balena.models.os
+	 */
+	const _getMaxSatisfyingVersion = function(
+		versionOrRange: keyof OsVersions,
```

Comment:
I think that this should be just `string` since it can hold a semver range as well.
Doing `| 'default'` would add some clarity for devs, but the resulting typings will just be widen to `string`.